beancount/parser/printer.py

```diff
@@ -349,7 +349,11 @@ def Note(self, entry, oss):
         self.write_metadata(entry.meta, oss)
 
     def Document(self, entry, oss):
-        oss.write('{e.date} document {e.account} "{e.filename}"'.format(e=entry))
+        oss.write(
+            '{e.date} document {e.account} "{filename}"'.format(
+                e=entry, filename=misc_utils.escape_string(entry.filename)
+            )
+        )
```

Comment:
🤔 #876 is basically just skip testing failed case, right?

Reply:
Wrong. The test is failing because it assumes something that is not true: that file paths do not contain characters that need to be escaped to be serialized as Beancount strings. This is not true in general on Windows. #876 constructs the test input data in a way that satisfies the assumption across platforms.

Reply:
This is just another way of saying "skip testing failed case"

Reply:
No it is not. "skip testing failed case" assumes that there is something wrong with the code being tested. In this case, as I explained, there is something wrong with the test. What it is done is to fix the test to test the right thing. If you still don't see the difference, I don't know how else to explain it.

Reply:
What you are missing is, "file paths do not contain characters that need to be escaped" is the failed case.

By removing `\r` `\n` and `\` testing, https://github.com/beancount/beancount/pull/876 effectively skip this case.

Reply:
"there is something wrong with the code being tested"

This is obvirously true, for a beancount like this on windows:

```beancount
row = r'2014-06-08 document Assets:Account1 "./document.csv"'
```

printer will output something like this:

```beancount
2014-06-08 document Assets:Account1 "C:\Users\Trim21\proj\beancount\document.csv"
```

Reply:
off course there is something wrong with the code being tested

did you try this on windwos:

```beancount
from beancount import print_entries
from beancount.loader import load_string

row = r'2014-06-08 document Assets:Account1 "./document.csv"'

print(load_string(row)[0])

print_entries(load_string(row)[0])
```

what should be expected output is currently output `2014-06-08 document Assets:Account1 "C:\Users\Trim21\proj\beancount\document.csv"` right?

Reply:
> The assumption that file paths / narration do not contain characters that need to be escaped is simply wrong

The existing code expects that characters that need escaping are escaped in the data passed to `beancount.parser.printer.print_entry()`. This is how it works. This is how it was designed. It cannot be changed. As I have explained multiple times, this would be an incompatible change that would break existing code.

As I have written multiple times, if you want to change this, please propose this change in a new PR where the merits of breaking backward compatibility can be discussed. It has nothing to do with fixing the tests to run on Windows.

> By removing `\r` `\n` and `\` from testing, #876 the tests are fixed, but failed
> case like `print_entries(load_string(r'2014-06-08 document Assets:Account1 "./document.csv"')[0])` and `2014-06-10 * "Entry with escaped \"quotes\" \r \n"` is no longer tested.

Of course removing the newline and form feed characters from the test string they are not tested anymore. I suspect that you have not read the commit message of the commit that changes the test. Have you? What is not clear in it? Do you understand why the test currently succeeds on Unix platforms but fails on Windows?

Reply:
> I have explained multiple times

No one is talking about changing it anymore, even in this PR it's reverted, don't know why you stick to it.

This is about how we fixing the tests to run on Windows. and even you don't like how I describe it, it's "just skip testing failed case" like `@unittest.skipIf(sys.platform == "win32", reason="os.sep")` and consider they work as expected.

> The existing code expects that characters that need escaping are escaped in the data passed to beancount.parser.printer.print_entry(). This is how it works. 

This is clearly wrong.

https://github.com/beancount/beancount/blob/2a455c7987c8ae81bba761f5da3d74e4711d9e8e/beancount/parser/printer.py#L435